server: fix the tenant server error handling

knz opened the pull request:
Needed for #99958.
Fixes  #97661.
Fixes #98868.
Epic: CRDB-23559
First commit from #100579.

Prior to this patch, if an error occurred during the initialization or
startup of a secondary tenant server, the initialization would leak
state into the stopper defined for that tenant. Generally, reusing
a stopper across server startup failures is not safe (and API
violation).

This patch fixes it by decoupling the intermediate stopper used for
orchestration from the one used per tenant server.

knz committed:
server: unexport some functions

Release note: None

knz committed:
server: save a reference to BaseConfig in SQLServerWrapper

Release note: None

knz committed:
server: extend the `onDemandServer` interface

ahead of splitting "new" vs "start" during construction.

Release note: None

knz committed:
server: lift `reportTenantInfo` into `*SQLServerWrapper`

Release note: None

knz committed:
server: separate "new" from "start" for tenant servers

This peels the call to "start" from the `newTenantServer`
interface and pulls it into the orchestration retry loop.

This change also incidentally reveals an earlier misdesign: we are
calling `newTenantServer` _then_ `start` in the same retry loop. If
`new` succeeds but `start` fails, the next retry will call
`newTenantServer` again *with the same stopper*, which will leak
closers from the previous call to `new`.

Release note: None

knz committed:
server: fix the tenant server error handling

Prior to this patch, if an error occurred during the initialization or
startup of a secondary tenant server, the initialization would leak
state into the stopper defined for that tenant. Generally, reusing
a stopper across server startup failures is not safe (and API
violation).

This patch fixes it by decoupling the intermediate stopper used for
orchestration from the one used per tenant server.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100436)
<!-- Reviewable:end -->

knz committed:
serverccl: simplify TestServerStartupGuardrails

Prior to this patch, the test was not cleaning up its server stopper
reliably at the end of each sub-test. This patch fixes it.

Release note: None

knz committed:
serverccl: make TestServerControllerMultiNodeTenantStartup faster

Release note: None

knz committed:
serverccl: clarify the progress inside TestServerControllerMultiNodeTenantStartup

Release note: None

knz commented:
Encountering this issue while troubleshooting TestServerControllerMultiNodeTenantStartup. Will come back to this later.

knz commented:
Adding this diff to help with #100578:

```diff
@@ -73,6 +74,12 @@ func (u Updater) update(ctx context.Context, useReadLock bool, updateFn UpdateFn
        ctx, sp := tracing.ChildSpan(ctx, "update-job")
        defer sp.Finish()

+       // Disable DistSQL query distribution. This ensures that the job operations do not
+       // require SQL servers to be ready on other nodes.
+       prevMode := u.txn.SessionData().DistSQLMode
+       defer func(prevMode sessiondatapb.DistSQLExecMode) { u.txn.SessionData().DistSQLMode = prevMode }(prevMode)
+       u.txn.SessionData().DistSQLMode = sessiondatapb.DistSQLOff
+
```
